src/Interpreters/TraceCollector.cpp

```diff
@@ -29,37 +29,46 @@ TraceCollector::TraceCollector(std::shared_ptr<TraceLog> trace_log_)
     thread = ThreadFromGlobalPool(&TraceCollector::run, this);
 }
 
+void TraceCollector::tryClosePipe()
+{
+    try
+    {
+        TraceSender::pipe.close();
+    }
+    catch (...)
+    {
+        tryLogCurrentException("TraceCollector");
+    }
+}
 
 TraceCollector::~TraceCollector()
 {
     try
     {
-        if (!thread.joinable())
-            LOG_ERROR(&Poco::Logger::get("TraceCollector"), "TraceCollector thread is malformed and cannot be joined");
+        if (thread.joinable())
+        {
+            /** Sends TraceCollector stop message
+            *
+            * Each sequence of data for TraceCollector thread starts with a boolean flag.
+            * If this flag is true, TraceCollector must stop reading trace_pipe and exit.
+            * This function sends flag with a true value to stop TraceCollector gracefully.
+            */
+            WriteBufferFromFileDescriptor out(TraceSender::pipe.fds_rw[1]);
+            writeChar(true, out);
+            out.next();
+        }
```

Comment:
@antonio2368 what's your opinion on this piece? I'm not sure why we need to check joinability on sending close message... probably send it regardless, and check joinability below - where we actually want to join? I'm not sure why it was done this way, maybe for reason...

Reply:
this part is only for stopping the thread (sending the flag to the thread). If the thread is stopped there is no need for it. Also, pipe could be invalid in this case so I assume we don't want to try sending anything through it.